libtenzir/builtins/operators/print.cpp

```diff
@@ -0,0 +1,198 @@
+        .docs("https://docs.tenzir.com/operators/print")
+        .throw_();
+    }
+    TENZIR_ASSERT(printer);
+    printer_ = printer->parse_printer(p);
+    TENZIR_ASSERT(printer_);
+  }
+
+  auto
+  operator()(generator<table_slice> input, operator_control_plane& ctrl) const
+    -> generator<table_slice> {
+    for (auto&& slice : input) {
+      if (slice.rows() == 0) {
+        co_yield {};
+        continue;
+      }
+      auto target_index
+        = slice.schema().resolve_key_or_concept_once(input_.inner);
+      if (not target_index) {
+        co_return diagnostic::error("could not resolve `{}` for schema `{}`",
+                                    input_.inner, slice.schema())
+          .primary(input_.source)
+          .emit(ctrl.diagnostics());
+      }
+      auto transform = [&](struct record_type::field field,
+                           std::shared_ptr<arrow::Array> array)
+        -> std::vector<
+          std::pair<struct record_type::field, std::shared_ptr<arrow::Array>>> {
+        if (not caf::holds_alternative<record_type>(field.type)) {
+          diagnostic::error("field {} is not of type record", field.name)
+            .primary(input_.source)
+            .throw_();
+        }
+        field.type = type{"dummy_head", field.type};
+        auto rb = arrow::RecordBatch::Make(
+          field.type.to_arrow_schema(), array->length(),
+          static_cast<const arrow::StructArray&>(*array).Flatten().ValueOrDie());
+        auto row = table_slice{rb, field.type};
+        auto builder = series_builder();
+        auto printer_instance = printer_->instantiate(field.type, ctrl);
+        auto row_as_slice = to_record_batch(row);
+        auto chunks = std::vector<chunk_ptr>{};
+        std::ranges::copy(printer_instance->get()->process(row),
+                          std::back_inserter(chunks));
+        std::ranges::copy(printer_instance->get()->finish(),
+                          std::back_inserter(chunks));
+        if (chunks.empty()) {
+          return {};
+        }
+        if (chunks.size() == 1) {
+          const auto* data = reinterpret_cast<const char*>(chunks[0]->data());
+          builder.data(std::string_view(data, data + chunks[0]->size()));
+        } else {
+          auto data = std::vector<std::byte>{};
+          for (auto&& chunk : chunks) {
+            data.insert(data.end(), chunk->begin(), chunk->end());
+          }
+          const auto* data_as_char = reinterpret_cast<const char*>(data.data());
+          builder.data(
+            std::string_view(data_as_char, data_as_char + data.size()));
+        }
+        auto series = builder.finish_assert_one_array();
+        auto check_utf_8
+          = arrow::compute::CallFunction("utf8_is_printable", {series.array});
+        if (not check_utf_8.ok()) {
+          diagnostic::error("{}. {} does not print UTF8. ", printer_->name(),
+                            check_utf_8.status().ToString())
+            .throw_();
+        }
```

Comment:
This is too strict, we want to allow all UTF8 results. We should check before we add to the builder.

Reply:
Consider experimenting with an API change like this that we can check upfront:

```diff
M libtenzir/include/tenzir/plugin.hpp
@@ -454,6 +454,10 @@ public:
   /// Returns whether the printer allows for joining output streams into a
   /// single saver.
   virtual auto allows_joining() const -> bool = 0;
+
+  /// Returns whether the printer has a text-based output formatted that is
+  /// guaranteed to be (1) valid UTF8 and (2) supposed to be human-readable.
+  virtual auto text_output() const -> bool = 0;
 };
 
 /// @see operator_parser_plugin
```